src/bin/drag_and_drop_listbox.rs

```diff
@@ -0,0 +1,193 @@
+    }
+
+    window.show_all();
+}
+
+fn create_row(i: i32) -> ListBoxRow {
+    let list_box_row = ListBoxRow::new();
+    let box_ = Box::new(Orientation::Horizontal, 5);
+    box_.set_margin_start(10);
+    box_.set_margin_end(10);
+    list_box_row.add(&box_);
+
+    let event_box = EventBox::new();
+    let image = Image::new_from_icon_name("open-menu-symbolic", IconSize::Button);
+    event_box.add(&image);
+    box_.add(&event_box);
+
+    let label_text = format!("Row {}", i);
+    let label = Label::new(Some(label_text.as_str()));
+
+    box_.add(&label);
+
+    // Drag stuff
+    //
+    // Make the event box draggable. Triggered with left mouse button.
+    //
+    let drag_targets: [TargetEntry; 1] = [TargetEntry::new(
+        "GTK_LIST_BOX_ROW", // I don't know how this string is used
+        TargetFlags::SAME_APP,
+        0,
+    )];
+    event_box.drag_source_set(ModifierType::BUTTON1_MASK, &drag_targets, DragAction::MOVE);
+    // drag_begin will be called when we start dragging. drag_begin is where we call rendering the
+    // widget being dragged next to mouse cursor.
+    event_box.connect_drag_begin(drag_begin);
+    // We don't use this, but just to demonstrate that things can be done when the drag ends.
+    event_box.connect_drag_end(drag_end);
+    // See https://developer.gnome.org/gtkmm-tutorial/stable/sec-dnd-signals.html.en
+    // Called when a source is being dragged and the target calls get_data()
+    event_box.connect_drag_data_get(drag_data_get);
+    // Set the same box as a drag destination.
+    event_box.drag_dest_set(DestDefaults::ALL, &drag_targets, DragAction::MOVE);
+    // Called when get_data() returns
+    event_box.connect_drag_data_received(drag_data_received);
+
+    list_box_row
+}
+
+fn drag_begin<W: IsA<Widget>>(w: &W, ctx: &DragContext) {
+    println!("drag_begin({:?})", w);
+
+    // Get containing ListBoxRow and render it next to the cursor while being dragged
+    match w.get_ancestor(ListBoxRow::static_type()) {
+        None => {
+            println!("Can't get ancestor");
+        }
+        Some(ancestor) => {
+            let alloc = ancestor.get_allocation();
+            let surface: cairo::ImageSurface =
+                cairo::ImageSurface::create(cairo::Format::ARgb32, alloc.width, alloc.height)
+                    .unwrap();
+            let surface_: &cairo::Surface = surface.as_ref();
+
+            {
+                ancestor.get_style_context().add_class("drag-icon");
+                let surface_ctx = cairo::Context::new(surface_);
+                ancestor.draw(&surface_ctx);
+                ancestor.get_style_context().remove_class("drag-icon");
+            }
+
+            let (x, y) = w.translate_coordinates(&ancestor, 0, 0).unwrap();
+            surface.set_device_offset(-x as f64, -y as f64);
+            ctx.drag_set_icon_surface(surface_);
+        }
+    }
+}
+
+fn drag_end<W: IsA<Widget>>(w: &W, _ctx: &DragContext) {
+    println!("drag_end({:?})", w);
```

Comment:
Does the `Debug` impl of the `DragContext` print anything meaningful apart from the pointer? Maybe you want to print some other meaningful values here then :)

Reply:
This line prints `drag_end(EventBox { inner: ObjectRef(Shared { inner: 0x55b9f31ade20, borrowed: true }), type: GtkEventBox })`.

I used prints for debugging, I'll just remove them now.